### DISCUSSION TITLE
How to update an exists field value?

### DISCUSSION THREAD

**Discussion by axiaoxin** [uber-go/zap#1240] [Q&A] [answered]
Such as a logger like: `logger = logger.With(zap.String("myfield", "v1"))`
Is there any way to update the myfield's value?

**Comment by prashantv** [comment, marked as answer]
No, there isn't a way to update existing fields, as zap encodes fields as they're added to the logger.

You can add the same field multiple times (#81), but that may lead to undefined behaviour, so I'd recommend using a separate log field.

**Comment by sc-atompower** [comment]
what if the value is a function that returns something or if the value is a pointer to a struct field?

**Reply by abhinav** [reply]
Hi, could you clarify? Zap cannot log functions. It will just print their addresses.
For pointers to primitive types, Zap will use `*p` field constructors (e.g. [Stringp](https://pkg.go.dev/go.uber.org/zap#Stringp)), which will log the value if non-nil, and `null` otherwise.

**Reply by sc-atompower** [reply]
Hey there. Right but I noticed you could pass a called/immediately invoked method so I was curious if this would log the value as it was updated. 
We see lots of power in our applications with the logging of values that change over time so for example:

```
		zap.String("status", func() string {
			return string(charger.Status)
		}())
```
as the charger status changes over time it would be nice to have the value of status at the time of logging reflected.

**Reply by sc-atompower** [reply]
Or would this method be invoked at the time of registration and never again.

**Reply by abhinav** [reply]
Yes, Go will invoke the anonymous function right away because of the `()` at the end, and Zap will only see the final value.

Basically, this:
```go
zap.String("status", func() string { return charger.Status }())
```

Is roughly the same as,

```go
x := "status"
y := charger.Status
zap.String(x, y)
```

I think maybe what you need is the `Lazy` field constructor proposed in https://github.com/uber-go/zap/pull/1041#issuecomment-1104664819. If you think that'll solve your problem, we can discuss it more in a new feature request.
